pkg/maintainer/spv/spv.go

```diff
@@ -59,8 +63,271 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
 
-	// TODO: Implement. For now, just wait.
-	<-ctx.Done()
+	for {
+		if err := sm.proveDepositSweepTransactions(); err != nil {
+			return fmt.Errorf(
+				"error while proving deposit sweep transactions: [%v]",
+				err,
+			)
+		}
+
+		// TODO: Add proving of other type of SPV transactions: redemption
+		// transactions, moving funds transaction, etc.
+
+		select {
+		case <-time.After(sm.config.IdleBackOffTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (sm *spvMaintainer) proveDepositSweepTransactions() error {
+	depositSweepTransactions, err := sm.getUnprovenDepositSweepTransactions()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get unproven deposit sweep transactions: [%v]",
+			err,
+		)
+	}
+
+	// TODO: Consider handling a situation in which the block headers in the
+	//       proof span multiple Bitcoin difficulty epochs.
+	requiredConfirmations, err := sm.chain.TxProofDifficultyFactor()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get transaction proof difficulty factor: [%v]",
+			err,
+		)
+	}
+
+	for _, transaction := range depositSweepTransactions {
+		_, proof, err := bitcoin.AssembleSpvProof(
+			transaction.Hash(),
+			uint(requiredConfirmations.Uint64()),
+			sm.btcChain,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to assemble SPV proof: [%v]", err)
+		}
+
+		mainUTXO, vault, err := parseTransactionInputs(
+			sm.btcChain,
+			sm.chain,
+			transaction,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"error while parsing transaction inputs: [%v]",
+				err,
+			)
+		}
+
+		if err := sm.chain.SubmitDepositSweepProofWithReimbursement(
+			transaction,
+			proof,
+			mainUTXO,
+			vault,
+		); err != nil {
+			return fmt.Errorf(
+				"failed to submit deposit sweep proof with reimbursement: [%v]",
+				err,
+			)
+		}
+	}
 
 	return nil
 }
+
+func (sm *spvMaintainer) getUnprovenDepositSweepTransactions() (
+	[]*bitcoin.Transaction,
+	error,
+) {
+	// TODO: Limit how far in the past we are looking for the events.
+	//       Possibly store latest checked height in memory or file.
+	depositSweepTransactionProposals, err :=
+		sm.chain.PastDepositSweepProposalSubmittedEvents(nil)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past deposit sweep proposal submitted events: [%v]",
+			err,
+		)
+	}
+
+	// There will often be multiple events emitted for a single wallet. Prepare
+	// a list of unique wallet public key hashes.
+	walletPublicKeyHashes := uniqueWalletPublicKeyHashes(
+		depositSweepTransactionProposals,
+	)
+
+	unprovenDepositSweepTransactions := []*bitcoin.Transaction{}
+
+	for _, walletPublicKeyHash := range walletPublicKeyHashes {
+		// TODO: Should we check the wallet's state before attempting to submit
+		//       the deposit sweep proof?
+		// TODO: Think what the limit of transactions retrieved should be.
+		walletTransactions, err := sm.btcChain.GetTransactionsForPublicKeyHash(
+			walletPublicKeyHash,
+			5,
```

Comment:
In theory, the wallet should not execute the next action if the previous one was not yet completed. I think checking 5 as a default is fine, but I would also allow this value to be passed as a parameter.

Reply:
Oh, wait. There is a chance the wallet performed e.g. 20 redemptions after sweeping. This function fetches all Bitcoin transactions no matter the type.

Assuming we execute sweep every 48 hours and redemption every 3 hours, it seems that the default value should be closer to 20.

Reply:
Done in 45bb022c94708fe000003e3ea54112c81f967cab.
The default value for the parameter is `20`.